File: streampark-flink/streampark-flink-client/streampark-flink-client-core/src/main/scala/org/apache/streampark/flink/client/impl/YarnApplicationClient.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.streampark.flink.client.impl

import org.apache.streampark.common.conf.Workspace
import org.apache.streampark.common.constants.Constants
import org.apache.streampark.common.enums.FlinkJobType
import org.apache.streampark.common.fs.FsOperator
import org.apache.streampark.common.util.{AssertUtils, FileUtils, HdfsUtils}
import org.apache.streampark.common.util.Implicits._
import org.apache.streampark.flink.client.`trait`.YarnClientTrait
import org.apache.streampark.flink.client.bean._
import org.apache.streampark.flink.packer.pipeline.ShadedBuildResponse

import org.apache.flink.client.deployment.application.ApplicationConfiguration
import org.apache.flink.configuration._
import org.apache.flink.python.PythonOptions
import org.apache.flink.yarn.YarnClusterDescriptor
import org.apache.flink.yarn.configuration.YarnConfigOptions
import org.apache.hadoop.yarn.api.records.ApplicationId

import java.util
import java.util.Collections

import scala.collection.mutable.ListBuffer

/** yarn application mode submit */
object YarnApplicationClient extends YarnClientTrait {

  private[this] lazy val workspace = Workspace.remote

  override def setConfig(submitRequest: SubmitRequest, flinkConfig: Configuration): Unit = {
    super.setConfig(submitRequest, flinkConfig)
    val providedLibs = {
      val array = ListBuffer(
        submitRequest.hdfsWorkspace.flinkLib,
        submitRequest.hdfsWorkspace.flinkPlugins,
        submitRequest.hdfsWorkspace.appJars)
      submitRequest.jobType match {
        case FlinkJobType.FLINK_SQL =>
          array += s"${workspace.APP_SHIMS}/flink-${submitRequest.flinkVersion.majorVersion}"
          val jobLib = s"${workspace.APP_WORKSPACE}/${submitRequest.id}/lib"
          if (HdfsUtils.exists(jobLib)) {
            array += jobLib
          }
        case _ =>
      }
      array.toList
    }

    flinkConfig
      // yarn.provided.lib.dirs
      .safeSet(YarnConfigOptions.PROVIDED_LIB_DIRS, providedLibs.asJava)
      // flinkDistJar
      .safeSet(YarnConfigOptions.FLINK_DIST_JAR, submitRequest.hdfsWorkspace.flinkDistJar)
      // pipeline.jars
      .safeSet(
        PipelineOptions.JARS,
        Collections.singletonList(
          submitRequest.buildResult
            .asInstanceOf[ShadedBuildResponse]
            .shadedJarPath))
      // yarn application name
      .safeSet(YarnConfigOptions.APPLICATION_NAME, submitRequest.effectiveAppName)
      // yarn application Type
      .safeSet(YarnConfigOptions.APPLICATION_TYPE, submitRequest.applicationType.getName)

    if (submitRequest.jobType == FlinkJobType.PYFLINK) {
      val pyVenv: String = workspace.APP_PYTHON_VENV
      AssertUtils.required(FsOperator.hdfs.exists(pyVenv), s"$pyVenv File does not exist")

      val localLib: String =
        s"${Workspace.local.APP_WORKSPACE}/${submitRequest.id}/lib"
      if (FileUtils.exists(localLib) && FileUtils.directoryNotBlank(localLib)) {
        flinkConfig.safeSet(PipelineOptions.JARS, util.Arrays.asList(localLib))
      }

      // yarn.ship-files
      val shipFiles = new util.ArrayList[String]()
      shipFiles.add(submitRequest.userJarFile.getParentFile.getAbsolutePath)

      flinkConfig
        .safeSet(YarnConfigOptions.SHIP_FILES, shipFiles)
        // python.files
        .safeSet(PythonOptions.PYTHON_FILES, submitRequest.userJarFile.getParentFile.getName)
        // python.archives
        .safeSet(PythonOptions.PYTHON_ARCHIVES, pyVenv)
        // python.client.executable
        .safeSet(PythonOptions.PYTHON_CLIENT_EXECUTABLE, Constants.PYTHON_EXECUTABLE)
        // python.executable
        .safeSet(PythonOptions.PYTHON_EXECUTABLE, Constants.PYTHON_EXECUTABLE)

      val args: util.List[String] =
        flinkConfig.get(ApplicationConfiguration.APPLICATION_ARGS)
      // Caused by: java.lang.UnsupportedOperationException
      val argsList: util.ArrayList[String] = new util.ArrayList[String](args)
      argsList.add("-pym")
      argsList.add(submitRequest.userJarFile.getName.dropRight(Constants.PYTHON_SUFFIX.length))
      flinkConfig.safeSet(ApplicationConfiguration.APPLICATION_ARGS, argsList)
    }

    logInfo(s"""
               |------------------------------------------------------------------
               |Effective submit configuration: $flinkConfig
               |------------------------------------------------------------------
               |""".stripMargin)
  }

  override def doSubmit(
      submitRequest: SubmitRequest,
      flinkConfig: Configuration): SubmitResponse = {
    val (clusterSpecification, clusterDescriptor: YarnClusterDescriptor) = getYarnClusterDeployDescriptor(flinkConfig, submitRequest.hadoopUser)
    logInfo(s"""
               |------------------------<<specification>>-------------------------
               |$clusterSpecification
               |------------------------------------------------------------------
               |""".stripMargin)

    val applicationConfiguration =
      ApplicationConfiguration.fromConfiguration(flinkConfig)
    var applicationId: ApplicationId = null
    var jobManagerUrl: String = null
    val clusterClient = clusterDescriptor
      .deployApplicationCluster(clusterSpecification, applicationConfiguration)
      .getClusterClient
    applicationId = clusterClient.getClusterId
    jobManagerUrl = clusterClient.getWebInterfaceURL
    logInfo(s"""
               |-------------------------<<applicationId>>------------------------
               |Flink Job Started: applicationId: $applicationId
               |__________________________________________________________________
               |""".stripMargin)

    val resp =
      SubmitResponse(applicationId.toString, flinkConfig.toMap, jobManagerUrl = jobManagerUrl)
    closeSubmit(submitRequest, clusterClient, clusterDescriptor)
    resp
  }

}

File: streampark-flink/streampark-flink-client/streampark-flink-client-core/src/main/java/org/apache/streampark/flink/client/impl/YarnApplicationClient.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.streampark.flink.client.impl;

import org.apache.streampark.common.configuration.Constants;
import org.apache.streampark.common.configuration.Workspace;
import org.apache.streampark.common.enums.FlinkJobType;
import org.apache.streampark.common.fs.FsOperator;
import org.apache.streampark.common.util.AssertUtils;
import org.apache.streampark.common.util.FileUtils;
import org.apache.streampark.common.util.HdfsUtils;
import org.apache.streampark.common.util.Tuple2;
import org.apache.streampark.flink.client.bean.RemoteWorkspace;
import org.apache.streampark.flink.client.bean.ResolvedSubmitRequest;
import org.apache.streampark.flink.client.request.SubmitRequest;
import org.apache.streampark.flink.client.response.SubmitResponse;
import org.apache.streampark.flink.packer.pipeline.ShadedBuildResponse;

import org.apache.flink.client.deployment.ClusterSpecification;
import org.apache.flink.client.deployment.application.ApplicationConfiguration;
import org.apache.flink.client.program.ClusterClient;
import org.apache.flink.configuration.Configuration;
import org.apache.flink.configuration.PipelineOptions;
import org.apache.flink.python.PythonOptions;
import org.apache.flink.util.FlinkException;
import org.apache.flink.yarn.YarnClusterDescriptor;
import org.apache.flink.yarn.configuration.YarnConfigOptions;
import org.apache.hadoop.yarn.api.records.ApplicationId;

import java.io.File;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;

/** Submits Flink jobs to application-scoped YARN clusters. */
public final class YarnApplicationClient extends AbstractYarnClient {

    public static final YarnApplicationClient INSTANCE = new YarnApplicationClient();

    private static final Workspace WORKSPACE = Workspace.REMOTE;

    private YarnApplicationClient() {
    }

    /** Configures remote artifacts and PyFlink resources for YARN application mode. */
    @Override
    protected void setConfig(ResolvedSubmitRequest resolved, Configuration flinkConfig) {
        SubmitRequest submitRequest = resolved.request();
        super.setConfig(resolved, flinkConfig);

        List<String> providedLibs = new ArrayList<>();
        RemoteWorkspace hdfsWorkspace = RemoteWorkspace.resolve(submitRequest.flinkVersion());
        providedLibs.add(hdfsWorkspace.flinkLib());
        providedLibs.add(hdfsWorkspace.flinkPlugins());
        providedLibs.add(hdfsWorkspace.appJars());

        if (submitRequest.jobType() == FlinkJobType.FLINK_SQL) {
            providedLibs.add(
                WORKSPACE.shims + "/flink-" + submitRequest.flinkVersion().majorVersion());
            String jobLib = WORKSPACE.workspace + "/" + submitRequest.id() + "/lib";
            try {
                if (HdfsUtils.exists(jobLib)) {
                    providedLibs.add(jobLib);
                }
            } catch (Exception e) {
                throw new IllegalStateException("Failed to check Flink SQL job lib path: " + jobLib, e);
            }
        }

        flinkConfig.set(YarnConfigOptions.PROVIDED_LIB_DIRS, providedLibs);
        flinkConfig.set(YarnConfigOptions.FLINK_DIST_JAR, hdfsWorkspace.flinkDistJar());
        flinkConfig.set(
            PipelineOptions.JARS,
            Collections.singletonList(
                ((ShadedBuildResponse) submitRequest.buildResult()).shadedJarPath()));
        flinkConfig.set(
            YarnConfigOptions.APPLICATION_NAME,
            resolved.getJobName());
        flinkConfig.set(
            YarnConfigOptions.APPLICATION_TYPE,
            submitRequest.applicationType().getName());

        if (submitRequest.jobType() == FlinkJobType.PYFLINK) {
            File userJar = resolved.getUserJarFile();
            AssertUtils.required(userJar != null, "PyFlink job archive is missing");
            String pyVenv = WORKSPACE.pythonVenv;
            AssertUtils.required(FsOperator.hdfs().exists(pyVenv), pyVenv + " File does not exist");

            String localLib = Workspace.LOCAL.workspace + "/" + submitRequest.id() + "/lib";
            if (FileUtils.exists(localLib) && FileUtils.directoryNotBlank(localLib)) {
                flinkConfig.set(PipelineOptions.JARS, Arrays.asList(localLib));
            }

            ArrayList<String> shipFiles = new ArrayList<>();
            shipFiles.add(userJar.getParentFile().getAbsolutePath());

            flinkConfig.set(YarnConfigOptions.SHIP_FILES, shipFiles);
            flinkConfig.set(PythonOptions.PYTHON_FILES, userJar.getParentFile().getName());
            flinkConfig.set(PythonOptions.PYTHON_ARCHIVES, pyVenv);
            flinkConfig.set(PythonOptions.PYTHON_CLIENT_EXECUTABLE, Constants.PYTHON_EXECUTABLE);
            flinkConfig.set(PythonOptions.PYTHON_EXECUTABLE, Constants.PYTHON_EXECUTABLE);

            List<String> args = flinkConfig.get(ApplicationConfiguration.APPLICATION_ARGS);
            ArrayList<String> argsList = new ArrayList<>(args);
            argsList.add("-pym");
            argsList.add(
                userJar.getName()
                    .substring(0, userJar.getName().length() - Constants.PYTHON_SUFFIX.length()));
            flinkConfig.set(ApplicationConfiguration.APPLICATION_ARGS, argsList);
        }

        logEffectiveSubmitConfiguration(flinkConfig);
    }

    /** Deploys an application cluster and returns its YARN and REST identities. */
    @Override
    protected SubmitResponse doSubmit(
                                      ResolvedSubmitRequest resolved,
                                      Configuration flinkConfig) throws FlinkException {
        SubmitRequest submitRequest = resolved.request();
        return execute(
            () -> {
                Tuple2<ClusterSpecification, YarnClusterDescriptor> deployDescriptor =
                    getYarnClusterDeployDescriptor(flinkConfig, submitRequest.hadoopUser());
                ClusterSpecification clusterSpecification = deployDescriptor._1;
                YarnClusterDescriptor clusterDescriptor = deployDescriptor._2;
                logClusterSpecification(clusterSpecification);

                ClusterClient<ApplicationId> clusterClient = null;
                try {
                    ApplicationConfiguration applicationConfiguration =
                        ApplicationConfiguration.fromConfiguration(flinkConfig);
                    clusterClient =
                        clusterDescriptor
                            .deployApplicationCluster(
                                clusterSpecification, applicationConfiguration)
                            .getClusterClient();
                    ApplicationId applicationId = clusterClient.getClusterId();
                    logYarnJobStarted(applicationId);
                    return new SubmitResponse(
                        applicationId.toString(),
                        flinkConfig.toMap(),
                        "",
                        clusterClient.getWebInterfaceURL());
                } finally {
                    closeSubmissionResources(clusterClient, clusterDescriptor);
                }
            });
    }
}
